<a href="https://colab.research.google.com/github/santiagoahl/ml_/blob/main/entregable_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Entregable número 1 - Santiago Ahumada Lozano
---



# Exercise 1.2
We begun thinking the messages as the frecuency of occurrence of the keywords whose the perceptron will work with,  then our unknown target fuction will look like:

$$
f:X\rightarrow Y
$$
Where,
\begin{align*}
&f(x)=+1, \text{if the message $x$ is considered spam} \\
&f(x)=-1, \text{if not}
\end{align*}

a) We tought of some keywords that will end up with large positive weights in the perceptron and we found this examples:


1.   Take advantage
2.   Right know
3.Unique
4. For sale
5. Unforgettable

b)We tought about inputs that will end up in negative values of $Y$ and we found the following examples:


1.   Invitation
2.   Security alert
3. Reminder: ...
4.Recommended 
5.Pay 

c)The parameter of the vector $x$ in the perceptron model which directly make the target function decide to clasify as spam is the number of keywords considered as spam i.e. its associated weights. 


---



# Exercise 1.3

Remmember that we want to find the correct vector of weights $$\mathbf{w}=(w_{1},...,w_{n})$$ 
By using the formula 
$$\mathbf{w}(t+1)=\mathbf{w}(t)+y(t)\mathbf{x}(t)$$

a)Now we want to show that $ \mathbf{y}(t)\mathbf{w}^{T}(t)\mathbf{x}(t) < 0 $ . In fact, since $x(t)$ is missclasified by $\mathbf{w}(t)$, then $y(t)\neq sign(\mathbf{w}^{T}(t)\mathbf{x}(t))$ and therefore its product must be negative.

b)Also we have $y(t)\mathbf{w}^{T}(t+1)\mathbf{x}(t) > y(t)\mathbf{w}^{T}(t)\mathbf{x}(t)$, 

\begin{align*}
y(t) \mathbf{w}^{T}(t+1) \mathbf{x}(t) &= y(t)( \mathbf{w}^{T}(t)+y(t)\mathbf{x}^{T}(t))\mathbf{x}(t) \\
 &= y(t)(( \mathbf{w}^{T}(t)\mathbf{x}(t)+y(t)\mathbf{x}^{T}(t)\mathbf{x}(t)) \\
 &=y(t)\mathbf{w}^{T}(t)\mathbf{x}(t) + y^{2}(t)\mathbf{x}^{T}(t)\mathbf{x}(t) \\
 &= y(t)\mathbf{w}^{T}(t)\mathbf{x}(t) +  y^{2}(t)\sum_{i=1}^{N}x_i^{2} \\
 &> y(t)\mathbf{w}^{T}(t)\mathbf{x}(t)
\end{align*} 

c)We've observed that each term  $y(t)\mathbf{w}^{T}(t+1)\mathbf{x}(t)>y(t) \mathbf{w}^{T}(t)\mathbf{x}(t)$, then there must be some value of $t$ such that $ y(t)\mathbf{w}^{T}(t)\mathbf{x}(t) > 0 $ i.e. our $\mathbf{x}$ will end up correctly clasified by $\mathbf{w}$.

# Exercise 1.10

Here is an experiment that ilustrates the difference between a single bin and multiple bins. Run a computer simulation for flipping 1000 fair coins. Flip each independly 10 times. Let's focus on 3 coins as follows: $c_1$ is the first coin flipped, $c_{rand}$ is a coin picked at random and $c_{min}$ is the coin that had the minimum frequency of heads. Let $\nu_{1}, \nu_{rand}, \nu_{min}$ the fraction of heads you obtain for the respective three coins.

# Solution.


In [86]:
import numpy as np
import random

nExperiments = 100
sum_mu = 0
sum_nu_1 = 0
sum_nu_rand = 0 
sum_C_min = 0

for e in range(nExperiments):
  nCoins = 1000

  val_1=0
  val_rand = random.choice(range(1,1000))
  val_min = 0
  #print('val_rand=',val_rand)

  C_1 = 0
  C_rand = 0
  C_min = 10

  #print(C_rand)

  nTrials = 1000

  for i in range(nTrials):
    #print('coin #',i+1)
    #Toss a fair coin 10 times
    sampleBin = np.random.binomial(1,0.5,10)
    nSuccess = np.sum(sampleBin)
    if i == val_rand:
      C_rand = nSuccess
    if i == val_1 : 
      C_1 = nSuccess
    if nSuccess < C_min :
      C_min = nSuccess
      val_min = i+1
      #print(val_min)
  #print('Number of heads of this coin',i+1,' ', nSuccess, 'Minimum_heads= ', C_min)

    nu_1 = C_1/10
    nu_rand = C_rand/10

    mu_exp_act = (nu_1+nu_rand+nu_min)/3

    #print('\n The average value of those 3 means (In the actual experiment) is',mu_exp_act)
    
  sum_nu_1 += nu_1
  sum_nu_rand += nu_rand 
  sum_C_min += C_min
  sum_mu += mu_exp_act

nu_1 = sum_nu_1/nExperiments
nu_rand = sum_nu_rand/nExperiments
C_min = int(sum_C_min/nExperiments)
mu = sum_mu/nExperiments

print('Probability of getting 10 heads if you toast the first fair coin 10 times:', nu_1)
print('Probability of getting 10 heads if you toast a random fair coin 10 times:', nu_rand)
print('The expected minimum number of heads is:', C_min)



Probability of getting 10 heads if you toast the first fair coin 10 times: 0.494
Probability of getting 10 heads if you toast a random fair coin 10 times: 0.467
The expected minimum number of heads is: 0


The probability found of getting 10 heads if you toast the first fair coin 10 times: 0.49
Probability of getting 10 heads if you toast a random fair coin 10 times: 0.46
The expected minimum number of heads is: 0
